lib/rolodex/headers.ex

```diff
@@ -0,0 +1,127 @@
+defmodule Rolodex.Headers do
+  @moduledoc """
+  Exposes functions and macros for defining reusable headers in route doc
+  annotations or responses.
+
+  It exposes the following macros, which when used together will set up the headers:
+
+  - `headers/2` - for declaring the headers
+  - `header/3` - for declaring a single header for the set
+
+  It also exposes the following functions:
+
+  - `is_headers_module?/1` - determines if the provided item is a module that has
+  defined a reusable headers set
+  - `to_map/1` - serializes the headers module into a map
+  """
+
+  alias Rolodex.Field
+
+  defmacro __using__(_) do
+    quote do
+      import Rolodex.Headers, only: :macros
+    end
+  end
+
+  @doc """
+  Opens up the headers definition for the current module. Will name the headers
+  set and generate a list of header fields based on the macro calls within.
+
+  **Accepts**
+  - `name` - the headers name
+  - `block` - headers shape definition
+
+  ## Example
+
+      defmodule SimpleHeaders do
+        use Rolodex.Headers
+
+        headers "SimpleHeaders" do
+          header "X-Rate-Limited", :boolean
+          header "X-Per-Page", :integer, desc: "Number of items in the response"
+        end
+      end
+  """
+  defmacro headers(name, do: block) do
+    quote do
+      Module.register_attribute(__MODULE__, :headers, accumulate: true)
+
+      unquote(block)
+
+      def __headers__(:name), do: unquote(name)
+      def __headers__(:headers), do: Map.new(@headers, fn {id, opts} -> {id, Field.new(opts)} end)
+    end
+  end
+
+  @doc """
+  Sets a header field.
+
+  **Accepts**
+
+  - `identifier` - the header name
+  - `type` - the header field type
+  - `opts` (optional) - additional metadata. See `Field.new/1` for a list of
+  valid options.
+  """
+  defmacro header(identifier, type, opts \\ []) do
```

Comment:
Not sure where else to put this, and I might be missing some additional context for how you’ve laid this out elsewhere, but would there be a path forward here whereby we don’t introduce another macro and instead rely on the more generic ‘field/3’ macro to define these headers? 

I’m also wondering if we absolutely need the addition of the headers/2 macro too and we could rely on more generic objects to build these types?

I understand the need for having these as specific types in some places, but looking through this the structure is very consistent to other types. Introducing more DSLs means more API for users to learn — would be good to reduce that if possible.

Reply:
Yea, I hear what you're saying. I think there are two parts here: the DSL for defining a reusable "component" (i.e. header/schema/etc) and then the DSL for defining the content within.

I think there's value in having a distinct `headers/2` macro for containing the shared component definition. For both parsing and serialization purposes, we need to identify the "type" of component in play. We could do something like:

```elixir
component "MyHeaders", type: :headers do
  ...
end
```

But, that's just another form of:

```elixir
headers "MyHeaders do
  ...
end
```

I think you're right, though, about the `header/3` macro for defining each header field. It really could just lean on the `field` macro a la schema. I'm okay making that change.

Does that all make sense?

Reply:
@cjbell here's a pass at what I'm thinking of based on your comments https://github.com/Frameio/rolodex/pull/48/commits/8440b211cc490367c8ab2e2bb15067f1c8108cb3